### PR TITLE
Proposal for verion 6.0.0

### PR DESCRIPTION
**Pull request by r3hp1c**
BC BREAK!!
Changed using subscriber/listener to doctrine types. This does not force annotations or attributes and should allow any supported Doctrine mappings. (Probably including YAML for older Doctrine versions)
Should improve speed and memory usage.
Add tests and support for Symfony 7 and Doctrine 3. Test on PHP 8.3.
Made Halite only require dev (documentation to be updated to indicate you have to install specifically)
Namespace changes
This should close most of the open issues when done.

- [ ] Fix all checks for all versions
- [x] Fix Encrypt/Decrypt commands for Doctrine ORM version 3 (ClassMetadata::CHANGETRACKING_NOTIFY does not exist any longer)
- [ ] Update documentation
- [ ] Update demo test cases for the DateTime, JSON and array types.
- [ ] What else was missed like composer.json require doctrine/annotations when probably no longer needed
- [ ] See other TODO messages
- [ ] Test the commands for all the supported Encrypt Types
- [ ] Squash commits

### CODE REVIEW

**Comment by core23** [comment]
Just two ideas what should also be part of the next major version:
- drop support for php 7 (and even 8.0) because they are no longer supported and prohibit using new features like type hints
- move YAML service configs to php service configs to improve processing time and get rid of the symfony yaml component

**Comment by r3hp1c** [comment]
@core23 thanks for the suggestions. dropping php7 support is going to make the code cleaner and possibly remove most deprecation notices. Will look into the suggestion to move to php service configuration as well.

**Comment by r3hp1c** [comment]
Firstly, all the random commits were because the lowest tests did not work as on my dev VM.

@core23 I still need to look at replacing yaml service config with php service config.

@core23 and @Zombaya I do have a task list in the description that I still need to look at, but does this proposal for version 6.0.0 close almost all open issues and are there any other suggestions and things I need to look at before I just continue with the task list?

File: .travis.yml [hunk 1]
```diff
@@ -8,10 +8,10 @@ matrix:
   fast_finish: true
   include:
       # Minimum supported versions
-    - php: 7.2
```
Comment (core23) [inline .travis.yml:11]: IMHO we can remove travis CI as we have GitHub CI for years.

The project also seems dead https://app.travis-ci.com/DoctrineEncryptBundle/DoctrineEncryptBundle

Comment (Zombaya) [reply .travis.yml:11]: Good catch.

**Comment by Zombaya** [comment]
Hi @r3hp1c, 

I have not done a full-in-depth review just yet, as I think we should first handle/discuss some major design choices.

# Overview
## Moving to doctrine/types
I'm not opposed to this, I think is certainly something we can investigate.  I think we should make sure we do introduce regressions:
  * Make sure we do not trigger an update every time we fetch an entity from the database, since encrypting the same string twice using halite will result in a different output.
  * Make the fields of child-entities get properly decrypted (although since we are not doing weird doctrine-events anymore, I don't foresee any trouble).
  * Perhaps others? Those two are two who came to mind.

I would also investigate if we would be able to introduce in the current 5.x-version of our project so that we can start showing deprecation notices there if users are using the annotation/attributes and have users migrate to the new method before having to upgrade to the next major version (similarly as how symfony does it).

## Moving to seperate namespace
The new namespace is `DoctrineEncryptBundle`, while one would expect it to be `DoctrineEncryptBundle\DoctrineEncryptBundle`, given that in general you would expect `name-of-organisation/name-of-project`.

I'm aware this currently the same, but I foresee problems when we would ever create another project in the same organisation.  For example. we would want to give that the namespace `DoctrineEncryptBundle\SecondProject`, which is not possible since the whole `DoctrineEncryptBundle`-top-namespace is claimed by this project.

I would also would like for us to try to migrate to the new namespace in a non-breaking way using [the strategy as suggested by ChatGPT](https://chat.openai.com/share/d39c8420-4837-4c1e-8aa6-4562c4eb9658):

```php
namespace Library\Old;

/**
 * @deprecated Use \Organisation\New instead.
 */
class MyClass
{
    // Your existing class implementation
}

// Provide an alias for the old namespace
class_alias(\Library\Old\MyClass::class, \Organisation\New\MyClass::class);

// You can also define a function or use an autoloader to handle the aliasing dynamically.
```
(I'm myself always sceptical of what ChatGPT suggest, but I have seen this kind of strategy before and I think it can work.)

## Dropping support for php 7.x
I would personally prefer if we were to keep in sync with the maintained versions of symfony.

The oldest maintened version for symfony is [5.4](https://symfony.com/releases), which supports [php 7.2.5+](https://symfony.com/releases/5.4).

I admit dropping support for 7.2 would have benefits, but I feel that this is not something a library should do. 

For end-projects, I fully support and encourage dropping support for non-supported php-versions.

## Move YAML service configs to php service configs
This is certainly something that can be done, but I would do this in a separate PR.

## In general
I feel that this PR is trying to solve too many problems at once. I think we/you should try to split this PR up into multiple PR's, each trying to solve a single problem.

Especially if the problem is able to be solved in the current master branch without creating breaking changes, but with addition of deprecation notices if users are not doing things as the next major version will only support.

Perhaps we can enable the discussions-module for this project and start a discussion on ideas on what to develop next.

I have some ideas of my own I'd like to develop when I get to it and I feel that sharing those beforehand should help us in getting on the same page on how and when to develop things before going in head-first.

File: .travis.yml (hunk 1 above)
Comment (r3hp1c) [reply .travis.yml:11]: Thanks for noticing that. Will get that done.

**Comment by r3hp1c** [comment]
Thanks for all your suggestions @Zombaya 

I really just created this merge to ensure not to loose the code I was playing with inside a little play project I have where I test some things.  O yes, and to get some feedback on the idea.

I fully agree that we need to do everything correctly and piece by piece.

Noted on the namespace changes.

Noted on PHP 7.2 and fully understand, Probably good to support all the PHP versions that the maintained Symfony versions support.

The only thing I need to figure out really is how to do the decrypt command while supporting both Annotations, Attributes and Doctrine Types. 

I can ensure you we will not be doing too many updates and re-encrypts.  The part I struggled to most with was getting the decrypt command to actually work.  Seems that the Doctrine UOW's original value is set after the convertToPHPValue function from the type, so the comparison is never done on the encrypted value inside the UOW.

Once I get some time and thought of a way to get the Annotations, Attributes and Doctrine Types working together I will start with the real work to start the deprecations and new implementations in version 5.4.

**Comment by r3hp1c** [comment]
Closing this. I have made a backup of what I had and going to start adding in piece by piece

**Comment by Zombaya** [comment]
@r3hp1c , just to make you sure you are aware and to avoid us doing duplicate work: I'm working on adding a new namespace and configuration key in integr8rs/DoctrineEncryptBundle#8 .

A lot of progress has been made, I currently think that I only need to create a rector to upgrade existing projects, update the demo's and add documentation for users on how to upgrade.

**Comment by r3hp1c** [comment]
@Zombaya I did see. Thank you.

I was about to start with adding the support for Doctrine types with an option to specify if you want to use annotations, attributes or the types when I saw your first message.

I believe if there is an option to specify, I should be able to get the encrypt and decrypt commands I was struggling with working by splitting the processing path for each type.

I am currently holding off until I can make the changes on the new namespace you are working on.

**Comment by Zombaya** [comment]
I think you should be able to continue with your changes, ignoring my new namespace. The new namespace is in a different directory, so it should not give any merge-conflicts. 

There are two things you/we will need to do when the code of both PRs is merged:
1. add a child-class in the new namespace, but that should only be a small change
2. Modify the services-definition as that is now done in a php-class rather than in yaml-files. This should not pose too much of a problem.

If you are working on the commands, might I suggest moving the business logic from the commands to a separate service so we can add unit tests for that as well?

I'm aware you technically can unit-test commands, but it's a bit of a bother in my opinion.

**Pull request closed without merging**